src/lib/eina/eina_thread_win32.c

```diff
@@ -0,0 +1,390 @@
+ * if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "eina_types.h"
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "eina_config.h"
+#include "eina_array.h"
+#include "eina_thread.h"
+#include "eina_sched.h"
+#include "eina_main.h"
+
+#include <assert.h>
+#include <evil_private.h>
+#include <process.h>
+
+/*
+ * The underlying type of Eina_Thread
+ */
+struct Thread
+{
+   CRITICAL_SECTION cancel_lock; /* mutex to protect the cancel handle */
+   char name[16]; /* the thread name */
+   HANDLE handle; /* thread handle */
+   void *data; /* on entry, the thread function argument, on exit, the return value */
+   Eina_Thread_Cb fn; /* the thread function */
+   Eina_Array *cleanup_fn;
+   Eina_Array *cleanup_arg;
+   unsigned id; /* thread id */
+   Eina_Bool free_on_exit; /* free the structure when thread exit */
+   volatile Eina_Bool cancel; /* the cancel event handle */
+   volatile Eina_Bool cancellable; /* is cancel enabled? */
+};
+
+typedef struct Thread Thread_t;
+
+/*
+ * This TLS stores the Eina_Thread for the current thread
+ */
+static DWORD tls_thread_self = 0;
+
+static Thread_t main_thread = { 0 };
+
+/*
+ * If we alloc'ed the Thread_t in eina_thread_self, we set
+ * free_on_exit flag to true, we then free it here
+ */
+void
+free_thread(void)
+{
+   Thread_t *t = TlsGetValue(tls_thread_self);
+   if (t && t->free_on_exit)
+      free(t);
+}
+
+static unsigned
+thread_fn(void *arg)
+{
+   Thread_t *thr = arg;
+   TlsSetValue(tls_thread_self, thr);
+   thr->data = thr->fn(thr->data, (Eina_Thread) thr);
+
+   while (eina_array_count(thr->cleanup_fn))
+     {
+        Eina_Thread_Cleanup_Cb fn = eina_array_pop(thr->cleanup_fn);
+        void *p = eina_array_pop(thr->cleanup_arg);
+        fn(p);
+     }
+
+   return 0;
+}
+
+Eina_Thread
+eina_thread_self(void)
+{
+    Thread_t *self = TlsGetValue(tls_thread_self);
+    /*
+     * If self is NULL this means
+     * 1) This function was called before eina_thread_init
+     * 2) This thread wasn't created by eina_thread_create
+     *
+     * In either case we alloc a new Thread struct and return
+     * it.
+     */
+    if (!self)
+      {
+         self = calloc(1, sizeof(*self));
+         self->handle = GetCurrentThread();
+         self->id = GetCurrentThreadId();
+         self->free_on_exit = EINA_TRUE;
+         if (tls_thread_self)
+            TlsSetValue(tls_thread_self, self);
+      }
+    return (Eina_Thread) self;
+}
+
+Eina_Bool
+eina_thread_equal(Eina_Thread t1, Eina_Thread t2)
+{
+   return ((Thread_t *) t1)->id == ((Thread_t *) t2)->id;
+}
+
+Eina_Bool
+eina_thread_create(Eina_Thread *t, Eina_Thread_Priority prio,
+                   int affinity, Eina_Thread_Cb func, const void *data)
+{
+   Thread_t *thr = calloc(1, sizeof(Thread_t));
+   if (!thr)
+      return EINA_FALSE;
+
+   thr->data = (void *) data;
+   thr->fn = func;
+
+   thr->handle = (HANDLE) _beginthreadex(NULL, 0, thread_fn, thr, CREATE_SUSPENDED, &thr->id);
```

Comment:
We don't need any of the operations from `eina_internal_call` then? AFAIK Win32 threads aren't cancelable, so this part is fine I guess, but what about the cleanup_push/pop with debug operations?

Reply:
Yeah, debug info is always good :D

Reply:
> but what about the cleanup_push/pop with debug operations?

@JPTIZ what do you mean?

Reply:
@walac in `eina_thread_posix.c` we always wrapped the given function with `_eina_internal_call`, which itself called those `EINA_THREAD_CLEANUP_PUSH`/`EINA_THREAD_CLEANUP_POP` with some debugging functions.

Reply:
Push with free function (no idea why): https://github.com/expertisesolutions/efl/pull/209/files/75acf308286175d146310f47dfe944a297e1ffb3#diff-034d507343e6df83b23e5ca8aaf7c37fR143

Push with debug function: https://github.com/expertisesolutions/efl/pull/209/files/75acf308286175d146310f47dfe944a297e1ffb3#diff-034d507343e6df83b23e5ca8aaf7c37fR193
And both Pops: https://github.com/expertisesolutions/efl/pull/209/files/75acf308286175d146310f47dfe944a297e1ffb3#diff-034d507343e6df83b23e5ca8aaf7c37fR195

Reply:
Oh, ok, the `free` is explained here: https://github.com/expertisesolutions/efl/pull/209/files/75acf308286175d146310f47dfe944a297e1ffb3#diff-034d507343e6df83b23e5ca8aaf7c37fR233

Reply:
Ok, added!